osquery/tables/system/linux/disk_encryption.cpp

```diff
@@ -114,34 +148,40 @@ QueryData genFDEStatus(QueryContext& context) {
     return results;
   }
 
-  std::map<std::string, Row> encrypted_rows;
-
   bool runSelectAll(true);
-  QueryData block_devices;
+  std::vector<std::string> queried_devices;
+  std::map<std::string, Row> block_devices;
+  std::map<std::string, Row> encrypted_rows;
 
   if (auto constraint_it = context.constraints.find("name");
       constraint_it != context.constraints.end()) {
     const auto& constraints = constraint_it->second;
-    for (const auto& name : constraints.getAll(EQUALS)) {
+    for (const auto& device : constraints.getAll(EQUALS)) {
       runSelectAll = false;
 
-      auto data = SQL::selectAllFrom("block_devices", "name", EQUALS, name);
-      for (const auto& row : data) {
-        block_devices.push_back(row);
+      if (!block_devices.count(device)) {
+        queryBlockDevice(runSelectAll, device, block_devices);
+        queried_devices.push_back(device);
       }
     }
   }
 
   if (runSelectAll) {
-    block_devices = SQL::selectAllFrom("block_devices");
+    queryBlockDevice(runSelectAll, "", block_devices);
   }
 
-  for (const auto& row : block_devices) {
-    const auto name = (row.count("name") > 0) ? row.at("name") : "";
-    const auto uuid = (row.count("uuid") > 0) ? row.at("uuid") : "";
-    const auto parent_name = (row.count("parent") > 0 ? row.at("parent") : "");
-    genFDEStatusForBlockDevice(
-        name, uuid, parent_name, encrypted_rows, results);
+  // Generate and add an encryption row result for each queried block device.
+  for (const auto& pair : block_devices) {
+    if (!encrypted_rows.count(pair.first)) {
+      genFDEStatusForBlockDevice(
+          runSelectAll, pair.second, block_devices, encrypted_rows);
+    }
+
+    if (queried_devices.empty() ||
+        std::count(
+            queried_devices.begin(), queried_devices.end(), pair.first)) {
```

Comment:
I wonder if `boost::algorithm::contains` would help. But that might be a bigger bite than needed

Reply:
`queried_devices.find(pair.first) != queried_devices.end()` is more idiomatic.

Reply:
I would use `std::find(queried_devices.begin(), queried_devices.end(), pair.first) != queried_devices.end()` to stop at the first result, since we don't need the count (and it's less efficient in the successful case).

Reply:
Sounds good! I've replaced count with find.